templates/e2e-dspy-workflow/README.ipynb

```diff
@@ -0,0 +1,1074 @@
+      "        return sanitized_prediction\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(inspect.getsource(IntentClassificationModule))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lastly, we set up some the vanilla program we will use throughout the notebook."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from src import MODEL_PARAMETERS, LOCAL_API_PARAMETERS\n",
+    "vanilla_program = IntentClassificationModule(labels_in_use)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Note: Run above this to do all setup without launching any models\n",
+    "# This is useful if you have already collected data and want to start from finetuning or from evaluation"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We will be using a local VLLM instance to run the initial benchmarks and data collection."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Gathering training data and running the 70B Model\n",
+    "\n",
+    "\n",
+    "## Preparation\n",
+    "\n",
+    "Before running the 70B model:\n",
+    "1. Remember to set your HF_TOKEN and HF_HOME environment variables\n",
+    "2. Use the following command to start the 70B server:\n",
+    "\n",
+    "   ```\n",
+    "   vllm serve meta-llama/Meta-Llama-3.1-70B-Instruct --port 8000 --pipeline_parallel_size 2 --enable_prefix_caching --tensor_parallel_size 2\n",
+    "   ```\n",
+    "\n",
+    "## Parallelism Configuration\n",
+    "\n",
+    "We've chosen pipeline parallelism and tensor parallelism of 2 for the 70B model based on our current setup. Here's the reasoning:\n",
+    "\n",
+    "1. Model size: The 70B model has 30 parts of ~5 GB each (based on [HuggingFace documentation](https://huggingface.co/meta-llama/Meta-Llama-3.1-70B-Instruct/tree/main)).\n",
+    "   - Total size: 30 * 5 GB = 150 GB\n",
+    "\n",
+    "2. Available VRAM:\n",
+    "   - Our GPUs: 80 GB VRAM x 4 = 320 GB\n",
+    "   - Tensor parallelism: floor(320/150) = 2\n",
+    "   - Pipeline parallelism: floor(num_gpus/2) = 2\n",
+    "   - To use all 4 GPUs efficiently:\n",
+    "     - Pipeline parallel size: 2\n",
+    "     - Tensor parallelism: 2\n",
+    "\n",
+    "3. Alternative setup (8x24GB GPUs):\n",
+    "   - Pipeline parallel size: 1\n",
+    "   - Tensor parallelism: ceil(150/24) = 7\n",
+    "\n",
+    "This configuration allows us to run the 70B model efficiently across our available GPU resources.\n",
+    "\n",
+    "Note that on Anyscale, you CANNOT download a 70B model without changing HF_HOME on most machines. The folder `/mnt/local_storage/' has enough space for a model download. It is not persisted across cluster restarts, but that is fine for model weights we don't need to save."
```

Comment:
Nit; I would rephrase this to say "On Anyscale Workspaces, we recommend to set ...."

Reply:
but it will crash if you don't change it

Reply:
Yeah but the wording is so negative here. Templates are user-facing and so you can expect to be one of the first points of contact with the platform. I would recommend better wording: 

"Note that by default, the cache directory used by HuggingFace is in the home directory (`/home/ray` here). We'll use `/mnt/local_storage` here for downloading large model weight files..."